### PR TITLE
Update vllm to use latest upstream to support CPU

### PR DESCRIPTION
**Pull request by xwu99**
Signed-off-by: Wu, Xiaochang <xiaochang.wu@intel.com>

### CODE REVIEW

**Comment by carsonwang** [comment]
Can we directly install vllm from pip as there is a v0.4.0.post1 release now?

**Comment by xwu99** [comment]
let me enable PR, it seems like the code is not in the watch list.

**Comment by carsonwang** [comment]
"vllm 0.4.0.post1+cpu requires torch==2.1.2+cpu, but you have torch 2.2.2+cpu which is incompatible." Please use a specific commit on vllm master branch which has upgraded to 2.2.1+cpu.

**Comment by xwu99** [comment]
> "vllm 0.4.0.post1+cpu requires torch==2.1.2+cpu, but you have torch 2.2.2+cpu which is incompatible." Please use a specific commit on vllm master branch which has upgraded to 2.2.1+cpu.

I don't think there is such commit. We need to fix torch version requirements.

**Comment by carsonwang** [comment]
> > "vllm 0.4.0.post1+cpu requires torch==2.1.2+cpu, but you have torch 2.2.2+cpu which is incompatible." Please use a specific commit on vllm master branch which has upgraded to 2.2.1+cpu.
> 
> I don't think there is such commit. We need to fix torch version requirements.

You can try the current vllm main branch to see if it can pass CI. If so, just use the current commit.

**Comment by carsonwang** [comment]
our requirement is torch>=2.2.0 so it should be compatible with vLLM's requirement (=2.2.1+cpu)

**Comment by xwu99** [comment]
> our requirement is torch>=2.2.0 so it should be compatible with vLLM's requirement (=2.2.1+cpu)

Updated version. but the "RuntimeError: Not support device type: cpu" is from vllm cpu, I am reporting it.

File: llm_on_ray/inference/vllm_predictor.py [hunk 1]
```diff
@@ -32,13 +33,17 @@ def __init__(self, infer_conf: InferenceConfig, max_num_seqs):
         model_config = model_desc.config
         dtype = "bfloat16" if infer_conf.vllm.precision == PRECISION_BF16 else "float32"
 
+        # Set environment variable VLLM_CPU_KVCACHE_SPACE to control the size of the CPU key-value cache.
+        # The current value is 40GB.
+        os.environ["VLLM_CPU_KVCACHE_SPACE"] = "40"
+
+        # Remove device=infer_conf.device here as CPU device type dispatch is not aligned in vLLM now
```
Comment (carsonwang) [inline llm_on_ray/inference/vllm_predictor.py:40]: This should have been fixed and need to pass the device, right?

**Comment by carsonwang** [comment]
Can you please update the user document how to use vllm as backend?

**Comment by xwu99** [comment]
> Can you please update the user document how to use vllm as backend?

Nothing ot update. Still follow docs/vllm.md as the setup is thru an install script.

File: llm_on_ray/inference/vllm_predictor.py [hunk 2]
```diff
@@ -25,20 +26,25 @@
...
     def __init__(self, infer_conf: InferenceConfig, max_num_seqs):
         super().__init__(infer_conf)
 
         model_desc = infer_conf.model_description
         model_config = model_desc.config
         dtype = "bfloat16" if infer_conf.vllm.precision == PRECISION_BF16 else "float32"
 
+        # Set environment variable VLLM_CPU_KVCACHE_SPACE to control the size of the CPU key-value cache.
+        # The default value is 40GB.
+        os.environ["VLLM_CPU_KVCACHE_SPACE"] = str(self.VLLM_CPU_KVCACHE_SPACE_DEFAULT)
```
Comment (carsonwang) [inline llm_on_ray/inference/vllm_predictor.py:40]: We should expose this config in the yaml file

Comment (xwu99) [reply llm_on_ray/inference/vllm_predictor.py:40]: Could we do this later as we will refactor the config file, 40GB works for most of the scenario now.

Comment (carsonwang) [reply llm_on_ray/inference/vllm_predictor.py:40]: yes, that will be fine.

**Comment by xwu99** [comment]
@carsonwang CI passed. Use huggingface-cli login to access gated llama2 model.
We can change to use a saved .cache/huggingface/token file that no need to login every time.

**Review by carsonwang — APPROVED** [review]

**Pull request merged**